File: solutions/0121_best-time-to-buy-and-sell-stock.py
# Problem: https://leetcode.com/problems/best-time-to-buy-and-sell-stock
# Runtime: 87 ms

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minSoFar = None
        bestSoFar = 0

        for p in prices:
            if minSoFar is None or p < minSoFar:
                minSoFar = p
            delta = p - minSoFar
            bestSoFar = max(delta, bestSoFar)
        
        return bestSoFar

File: solutions/0121_best-time-to-buy-and-sell-stock.java
// Problem: https://leetcode.com/problems/best-time-to-buy-and-sell-stock

class Solution {
    public int maxProfit(int[] prices) {
        int cheapest = prices[0];
        int maxProfit = 0;

        for (int i = 1; i < prices.length; i++) {
            int val = prices[i];

            if (val < cheapest) {
                cheapest = val;
            }
            if (val - cheapest > maxProfit) {
                maxProfit = val - cheapest;
            }
        }
        
        return maxProfit;
    }
}